### ISSUE TITLE
Array key type juggling

### ISSUE THREAD

**Issue by ShiraNai7** [vimeo/psalm#2241]
[PHP docs on array keys](https://www.php.net/manual/en/language.types.array.php):

> The key can either be an integer or a string.
>
> Additionally the following key casts will occur:
> 
> - Strings containing valid decimal integers, unless the number is preceded by a + sign, will be cast to the integer type. E.g. the key "8" will actually be stored under 8. On the other hand "08" will not be cast, as it isn't a valid decimal integer.

This means that even if I define my types as `array<string, some-type>` (or `iterable<string, some-type>`) and Psalm says the code is OK, I can still get integer keys during runtime if some of the strings happen to consist of digits.

Here is a pseudo-realistic example: https://psalm.dev/r/099b785b71

It will always crash with the following error:

> Uncaught TypeError: Argument 1 passed to ObjectCache::set() must be of the type string, int given

The fix in this case is to add a `string` cast when using `$key`:

```php
$this->set((string) $key, $value);
```

Because I'm aware of this issue, I'd use `array<array-key, some-type>` instead, but this means that `array<string, some-type>` might be "unsafe" to use in many cases. Is this something Psalm should detect/warn about?

**Comment by ShiraNai7** [comment]
Another issue is that passing `array-key` to a function that only accepts `string` produces `MixedArgumentTypeCoercion` (which is suppressed by `totallyTyped="false"`) instead of `PossiblyInvalidArgument`. I thought `array-key` meant `string|int`, not `mixed`?

https://psalm.dev/r/bff3415ffd

**Comment by muglug** [comment]
There's no way around these issues - PHP array key type coercion makes the language unsound, and it's beyond the scope of Psalm to fix this.
